### PR TITLE
A calculation function for StatefulSet maxUnavailable and some tests for it

### PR DESCRIPTION
**Pull request by FillZpp**
Signed-off-by: FillZpp <FillZpp.pub@gmail.com>

#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Add a calculation function for StatefulSet maxUnavailable and some tests for it.

#### Which issue(s) this PR fixes:

Fixes [comment](https://github.com/kubernetes/kubernetes/pull/110069#issuecomment-1140902579)

#### Special notes for your reviewer:

NONE

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/statefulset/stateful_set_utils_test.go [hunk 1]
```diff
@@ -937,3 +938,36 @@ func newStatefulSetWithLabels(replicas int, name string, uid types.UID, labels m
...
+		replicaCount           int
+		expectedMaxUnavailable int
+	}{
+		{maxUnavailable: nil, replicaCount: 10, expectedMaxUnavailable: 1},
+		{maxUnavailable: intOrStrP(intstr.FromInt(3)), replicaCount: 10, expectedMaxUnavailable: 3},
+		{maxUnavailable: intOrStrP(intstr.FromInt(3)), replicaCount: 0, expectedMaxUnavailable: 3},
+		{maxUnavailable: intOrStrP(intstr.FromString("10%")), replicaCount: 25, expectedMaxUnavailable: 2},
+		{maxUnavailable: intOrStrP(intstr.FromString("100%")), replicaCount: 5, expectedMaxUnavailable: 5},
+		{maxUnavailable: intOrStrP(intstr.FromString("10%")), replicaCount: 5, expectedMaxUnavailable: 1},
+		{maxUnavailable: intOrStrP(intstr.FromString("10%")), replicaCount: 5, expectedMaxUnavailable: 1},
```
Comment (atiratree) [inline pkg/controller/statefulset/stateful_set_utils_test.go:954]: duplicate test

File: pkg/controller/statefulset/stateful_set_utils_test.go [hunk 2]
```diff
@@ -937,3 +938,36 @@ func newStatefulSetWithLabels(replicas int, name string, uid types.UID, labels m
...
+		expectedMaxUnavailable int
+	}{
+		{maxUnavailable: nil, replicaCount: 10, expectedMaxUnavailable: 1},
+		{maxUnavailable: intOrStrP(intstr.FromInt(3)), replicaCount: 10, expectedMaxUnavailable: 3},
+		{maxUnavailable: intOrStrP(intstr.FromInt(3)), replicaCount: 0, expectedMaxUnavailable: 3},
+		{maxUnavailable: intOrStrP(intstr.FromString("10%")), replicaCount: 25, expectedMaxUnavailable: 2},
+		{maxUnavailable: intOrStrP(intstr.FromString("100%")), replicaCount: 5, expectedMaxUnavailable: 5},
+		{maxUnavailable: intOrStrP(intstr.FromString("50%")), replicaCount: 5, expectedMaxUnavailable: 2},
+		{maxUnavailable: intOrStrP(intstr.FromString("10%")), replicaCount: 5, expectedMaxUnavailable: 1},
+		{maxUnavailable: intOrStrP(intstr.FromString("1%")), replicaCount: 0, expectedMaxUnavailable: 1},
```
Comment (atiratree) [inline pkg/controller/statefulset/stateful_set_utils_test.go:955]: hm, I suppose it wouldn't hurt to also test the `0` and `0%` even with this comment in mind

```
// Note that API validation has already guaranteed the maxUnavailable field to be >1 if it is an integer
// or 0% < value <= 100% if it is a percentage, so we don't have to consider other cases.
```

Comment (FillZpp) [reply pkg/controller/statefulset/stateful_set_utils_test.go:957]: no pro, added :)

**Review by soltysh — APPROVED** [review]
/approve

**Pull request merged**